realtek: rtl93xx: add support for trapping management frames

sharadanand opened the pull request:
To properly support protocols like LLDP, EAPOL, and MSTP, the driver must configure the switch to trap their management frames to the CPU rather than forwarding them. This commit adds support for configuring these frame traps on RTL93xx devices.

Djfe commented:
I'm not sure what Test Formalities is complaining about. maybe a missing space between name and mail address in the "Signed-off-by" line?

0xharshal committed:
realtek: rtl93xx: Trap LLDP management frames

LLDP packets must be transmitted on a single port and trapped on a port of
a device which understands LLDP. It must not forward it to other ports to
avoid confusing neighbor information on connected devices.

Signed-off-by: Harshal Gohel <hg@simonwunderlich.de>
Signed-off-by: Sharadanand Karanjkar <sk@simonwunderlich.de>
Link: https://github.com/openwrt/openwrt/pull/19571
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

0xharshal committed:
realtek: dsa: Fix prefix for trapping functions

The functions to enable trapping of management frames are not RTL83xx
specific. It is more appropriate to use the more generic "rtldsa" prefix
for them.

Signed-off-by: Harshal Gohel <hg@simonwunderlich.de>
Signed-off-by: Sharadanand Karanjkar <sk@simonwunderlich.de>
Link: https://github.com/openwrt/openwrt/pull/19571
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

0xharshal committed:
realtek: rtl931x: Register support for trapping management frames

Driver needs to configure management frame actions
To support LLDP, EAPOL or MSTP, which needs to be trapped to the CPU
instead of being forwarded.

The function to implement the various management frame actions was already
present but not yet registered correctly.

Signed-off-by: Harshal Gohel <hg@simonwunderlich.de>
Signed-off-by: Sharadanand Karanjkar <sk@simonwunderlich.de>
Link: https://github.com/openwrt/openwrt/pull/19571
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

0xharshal committed:
realtek: rtl930x: Add support for trapping management frames

Driver needs to configure management frame actions
To support LLDP, EAPOL or MSTP, which needs to be trapped to the CPU
instead of being forwarded

Signed-off-by: Harshal Gohel <hg@simonwunderlich.de>
Signed-off-by: Sharadanand Karanjkar <sk@simonwunderlich.de>
Link: https://github.com/openwrt/openwrt/pull/19571
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

howels commented:
Is this ready for testing with LLDP?

ecsv reviewed (changes requested):
For other people reviewing the changes:

* https://svanheule.net/realtek/longan/register/rma_port_bpdu_ctrl
* https://svanheule.net/realtek/longan/register/rma_port_ptp_ctrl
* https://svanheule.net/realtek/longan/register/rma_port_lldp_ctrl
* https://svanheule.net/realtek/longan/register/rma_port_eapol_ctrl
* https://github.com/plappermaul/realtek-doc/blob/54589ff0afa70045abfdc71a3133aa55c76257bc/sources/rtk-xgs1210/system/include/private/drv/swcore/swcore_rtl9300.h#L6567
* https://github.com/plappermaul/realtek-doc/blob/54589ff0afa70045abfdc71a3133aa55c76257bc/sources/rtk-xgs1210/src/dal/longan/dal_longan_trap.c#L3263

ecsv commented:
> Is this ready for testing with LLDP?

What do you mean by testing? That LLDP is no longer forwarded to other ports but instead trapped to the CPU? After looking at the code, I would assume that this should work on RTL930x.

ecsv commented on target/linux/realtek/files-6.12/drivers/net/dsa/rtl83xx/rtl930x.c:
```diff
@@ -658,6 +658,76 @@ static void rtl930x_write_mcast_pmask(int idx, u64 portmask)
+		shift = (port % 10) * 3;
+		sw_w32_mask(GENMASK(shift + 2, shift), value << shift, reg);
+		break;
+	case GRATARP:
+		/* NOSUPPORT */
+		break;
```
```suggestion
	case GRATARP:
		reg = RTL930X_SPCL_TRAP_PORT_CTRL + (port / 16) * 4;
		shift = (port % 16) * 2;
		sw_w32_mask(GENMASK(shift + 1, shift), value << shift, reg);
		break;
```


```
#define RTL930X_SPCL_TRAP_PORT_CTRL		(0xA1A0)
```

* https://github.com/plappermaul/realtek-doc/blob/54589ff0afa70045abfdc71a3133aa55c76257bc/sources/rtk-xgs1210/src/dal/longan/dal_longan_trap.c#L3173C1-L3173C104
* https://svanheule.net/realtek/longan/register/spcl_trap_port_ctrl
* https://github.com/plappermaul/realtek-doc/blob/54589ff0afa70045abfdc71a3133aa55c76257bc/sources/rtk-xgs1210/system/include/private/drv/swcore/swcore_rtl9300.h#L4405C115-L4405C119

howels commented:
Tested on the Onti (clone of Xikestor SK8300X-8S) and while I do now see LLDP frames on the CPU port, there is also quite a lot of LLDP duplication occuring, which presents as the attached clients seeing frames from all active switch ports instead of just the attached interface:
```
-------------------------------------------------------------------------------
LLDP neighbors:
-------------------------------------------------------------------------------
Interface:    enp6s0, via: LLDP, RID: 303, Time: 0 day, 00:07:14
  Chassis:     
    ChassisID:    mac d0:aa:5f:00:1a:64
    SysName:      onti
    SysDescr:     OpenWrt SNAPSHOT @ onti
    MgmtIP:       A.B.C.D
    MgmtIface:    12
    MgmtIP:       A.B.C.D IPv6
    MgmtIface:    2
    Capability:   Bridge, on
    Capability:   Router, on
    Capability:   Wlan, off
    Capability:   Station, off
  Port:        
    PortID:       mac d0:aa:5f:00:1a:64
    PortDescr:    lan1
    TTL:          120
-------------------------------------------------------------------------------
Interface:    enp6s0, via: LLDP, RID: 303, Time: 0 day, 00:07:14
  Chassis:     
    ChassisID:    mac d0:aa:5f:00:1a:64
    SysName:      onti
    SysDescr:     OpenWrt SNAPSHOT @ onti
    MgmtIP:       A.B.C.D
    MgmtIface:    12
    MgmtIP:      A.B.C.D IPv6
    MgmtIface:    2
    Capability:   Bridge, on
    Capability:   Router, on
    Capability:   Wlan, off
    Capability:   Station, off
  Port:        
    PortID:       mac d2:aa:5f:00:1a:67
    PortDescr:    lan4
    TTL:          120
-------------------------------------------------------------------------------
Interface:    enp6s0, via: LLDP, RID: 303, Time: 0 day, 00:07:14
  Chassis:     
    ChassisID:    mac d0:aa:5f:00:1a:64
    SysName:      onti
    SysDescr:     OpenWrt SNAPSHOT @ onti
    MgmtIP:       A.B.C.D
    MgmtIface:    12
    MgmtIP:       A.B.C.D IPv6
    MgmtIface:    2
    Capability:   Bridge, on
    Capability:   Router, on
    Capability:   Wlan, off
    Capability:   Station, off
  Port:        
    PortID:       mac d2:aa:5f:00:1a:68
    PortDescr:    lan5
    TTL:          120
-------------------------------------------------------------------------------
```

howels commented:
Tagged VLAN traffic is accepted on the CPU port, but not forwarded to other tagged switch ports when this branch is used.

howels commented:
> > Tagged VLAN traffic is accepted on the CPU port, but not forwarded to other tagged switch ports when this branch is used.
> 
> @howels can't test this myself. But I find it a little bit odd (after I reviewed the code) because should not interact with the VLANs. But maybe we can understand this better: Is it triggered by the line
> 
> ```
> 	rtl83xx_setup_lldp_traps(priv);
> ```
> 
> or
> 
> ```
> 	rtl83xx_setup_bpdu_traps(priv);
> ```
> 
> Or by both.
> 
> Just remove these lines (1. the first one, 2. the second one, 3. both) in `rtl93xx_setup(), build a new images and test again.

I commented out the BPDU trap setup and tagged VLANs are working again:
```
	rtl83xx_setup_lldp_traps(priv);
	//rtl83xx_setup_bpdu_traps(priv);
```

ecsv commented:
> Tested on the Onti (clone of Xikestor SK8300X-8S) and while I do now see LLDP frames on the CPU port, there is also quite a lot of LLDP duplication occuring, which presents as the attached clients seeing frames from all active switch ports instead of just the attached interface:

So you attach a device on port LAN1 but it also sees LLDP traffic from LAN2. Yes, this is to be expected and not related to this tickets at all. To fix LLDP/STP sendout of such traffic, you need a patch like:


```diff
From: Issam Hamdi <ih@simonwunderlich.de>
Date: Fri, 28 Mar 2025 10:39:44 +0100
Subject: realtek: rtl93xx: change the fwd type in the tx tag to physical

During testing of LLDP and STP packets, we observed that when locally
generated multicast packets, such as LLDP or STP, were designated to be
sent on a specific port (e.g., lan1), they were also forwarded to other
ports.

To reproduce the issue:

* Set the interface in the lldpd uci config to "lan1" and restart the
  service.
* Connect two devices (e.g., PC) to the "lan1" and "lan2".
* Check the device connected to "lan2" and you will find it receive the
  lldp packets.

The solution:

* Is to change the 'FWD_TYPE' in the TX CPU TAG to specify the outgoing
  port.

Signed-off-by: Issam Hamdi <ih@simonwunderlich.de>

diff --git a/target/linux/realtek/files-6.12/drivers/net/ethernet/rtl838x_eth.c b/target/linux/realtek/files-6.12/drivers/net/ethernet/rtl838x_eth.c
index 7b2e6f6a15ab172cab7c57196ddc61462ba5b5b7..59bac6176ce21df4d517f2093d92490d7abc7143 100644
--- a/target/linux/realtek/files-6.12/drivers/net/ethernet/rtl838x_eth.c
+++ b/target/linux/realtek/files-6.12/drivers/net/ethernet/rtl838x_eth.c
@@ -170,7 +170,8 @@ static void rtl930x_create_tx_header(struct p_hdr *h, unsigned int dest_port,
 				     int prio, unsigned char *dest)
 {
 	h->cpu_tag[0] = 0x8000;  /* CPU tag marker */
-	h->cpu_tag[1] = h->cpu_tag[2] = 0;
+	h->cpu_tag[1] = 0x0100;  /* Set FWD_TYPE to 1 */
+	h->cpu_tag[2] = 0;
 	h->cpu_tag[3] = 0;
 	h->cpu_tag[4] = 0;
 	h->cpu_tag[5] = 0;
@@ -190,7 +191,8 @@ static void rtl931x_create_tx_header(struct p_hdr *h, unsigned int dest_port,
 				     int prio, unsigned char *dest)
 {
 	h->cpu_tag[0] = 0x8000;  /* CPU tag marker */
-	h->cpu_tag[1] = h->cpu_tag[2] = 0;
+	h->cpu_tag[1] = 0x0100;  /* Set FWD_TYPE to 1 */
+	h->cpu_tag[2] = 0;
 	h->cpu_tag[3] = 0;
 	h->cpu_tag[4] = h->cpu_tag[5] = h->cpu_tag[6] = h->cpu_tag[7] = 0;
 	if (dest_port >= 32) {
```

plappermaul reviewed (changes requested):
Thanks for your contribution. Please see comments.

plappermaul reviewed (approved):
Thanks.
